NorthstarDedicatedTest/dllmain.cpp

```diff
@@ -67,6 +77,96 @@ void WaitForDebugger(HMODULE baseAddress)
 			Sleep(100);
 	}
 }
+bool LoadPlugins() {
+
+	std::vector<fs::path> paths;
+
+	std::string pluginPath = GetNorthstarPrefix() + "/plugins";
+
+	// ensure dirs exist
+	fs::create_directories("plugins");
+
+	// get mod directories
+
+	for (auto const& entry : fs::recursive_directory_iterator(pluginPath))
+	{
+		if (fs::is_regular_file(entry) && entry.path().extension() == ".dll")
+			paths.emplace_back(entry.path().filename());
+	}
+
+	initGameState();
+	//spdlog::info("Loading the following DLLs in plugins folder:");
+	for (fs::path path : paths)
+	{
+		std::string pathstring = (pluginPath/ path).string();
+		std::wstring wpath = (pluginPath / path).wstring();
+		
+		LPCWSTR wpptr = wpath.c_str();
+		HMODULE datafile = LoadLibraryExW(wpptr, 0, LOAD_LIBRARY_AS_DATAFILE | LOAD_LIBRARY_AS_IMAGE_RESOURCE); // Load the DLL as a data file
+		if (datafile == NULL)
+		{
+			spdlog::info("Failed to load library {}. It appears the file does exist", pathstring);
+			continue;
+		}
+		HRSRC manifestResource = FindResourceW(datafile, MAKEINTRESOURCE(101), MAKEINTRESOURCE(RT_RCDATA));
+		
+		if (manifestResource == NULL)
+		{
+			spdlog::info("Could not find manifest for library {}", pathstring);
+			continue;
+		}
+		spdlog::info("Loading resource from library");
+		HGLOBAL myResourceData = LoadResource(datafile, manifestResource);
+		if (myResourceData == NULL)
+		{
+			spdlog::error("Failed to load resource from library");
+			continue;
+		}
+		int manifestSize = SizeofResource(datafile, manifestResource);
+		const char* manifestBinaryData = (const char*)LockResource(myResourceData);
+
+		std::string manifest = std::string(manifestBinaryData, 0, manifestSize);
+
+		rapidjson_document manifestJSON;
+		manifestJSON.Parse(manifest.c_str());
+
+		if (manifestJSON.HasParseError())
+		{
+			spdlog::error("Manifest for {} was invalid", pathstring);
+			continue;
+		}
+
+		if (!manifestJSON.HasMember("version"))
+		{
+			spdlog::error("{} does not have a version number in its manifest", pathstring);
+			continue;
+			//spdlog::info(manifestJSON["version"].GetString());
+		}
+
+		if (strcmp(manifestJSON["version"].GetString(), "1.0"))
+		{
+			spdlog::error("{} has an incompatible API version number in its manifest", pathstring);
+			continue;
+		}
+		// Passed all checks, going to actually load it now
+
+		HMODULE pluginLib = LoadLibraryW(wpptr); // Load the DLL as a data file
+		if (pluginLib == NULL)
+		{
+			spdlog::info("Failed to load library {}. It appears the file does exist", pathstring);
+			continue;
+		}
+		initPluginFuncPtr initPlugin = (initPluginFuncPtr)GetProcAddress(pluginLib, "initializePlugin");
+		if (initPlugin == 0)
+		{
+			spdlog::info("Library {} has no function initializePlugin", pathstring);
+			continue;
+		}
+		spdlog::info("Succesfully loaded {}", pathstring);
+		initPlugin(&gameState);
```

Comment:
For future extensibility, I think initPlugin should be refactored so it takes a function pointer to a function which gets an object by an ID (i.e., pass initPlugin a function pointer to a function like `void *getNorthstarObject(PLUGIN_OBJECT_TYPE_ID oid)`) (similar to how COM works).